server/sonar-server/src/main/java/org/sonar/server/authentication/JwtToken.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.authentication;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.ExpiredJwtException;
+import io.jsonwebtoken.JwtBuilder;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.SignatureException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import org.sonar.api.Startable;
+import org.sonar.api.config.Settings;
+import org.sonar.api.server.ServerSide;
+import org.sonar.api.utils.System2;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.core.util.UuidFactory;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * This class can be used to encode or decode a JWT token
+ */
+@ServerSide
+public class JwtToken implements Startable {
+
+  private static final Logger LOG = Loggers.get(JwtToken.class);
+
+  private static final String SECRET_KEY_PROPERTY = "sonar.secretKey";
+
+  private static final SignatureAlgorithm SIGNATURE_ALGORITHM = SignatureAlgorithm.HS256;
+
+  private final Settings settings;
+  private final System2 system2;
+  private final UuidFactory uuidFactory;
+
+  private SecretKey secretKey;
+
+  public JwtToken(Settings settings, System2 system2, UuidFactory uuidFactory) {
+    this.settings = settings;
+    this.system2 = system2;
+    this.uuidFactory = uuidFactory;
+  }
+
+  String encode(Jwt jwt) {
+    long now = system2.now();
+    JwtBuilder jwtBuilder = Jwts.builder()
+      .setId(uuidFactory.create())
+      .setSubject(jwt.getUserLogin())
+      .setIssuedAt(new Date(now))
+      .setExpiration(new Date(now + jwt.getExpirationTimeInSeconds() * 1000))
+      .signWith(SIGNATURE_ALGORITHM, secretKey);
+    for (Map.Entry<String, Object> entry : jwt.getProperties().entrySet()) {
+      jwtBuilder.claim(entry.getKey(), entry.getValue());
+    }
+    return jwtBuilder.compact();
+  }
+
+  Optional<Claims> decode(String token) {
```

Comment:
you should check that `start()` has been executed by doing: `Objects.requireNonNull(secretKey, "%s not started", getClass().getName())`